### PR TITLE
feat: add metrics and events for policy preemption

### PR DESCRIPTION
**Pull request by whitewindmills**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add metrics and events for policy preemption

**Which issue(s) this PR fixes**:
part of [feature](https://github.com/karmada-io/karmada/issues/3787)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/detector/preemption.go [hunk 1]
```diff
@@ -104,7 +107,18 @@ func (d *ResourceDetector) preemptPropagationPolicy(resourceTemplate *unstructur
 		return nil
 	}
 
-	if err := d.ClaimPolicyForObject(resourceTemplate, policy.Namespace, policy.Name); err != nil {
+	defer func() {
```
Comment (jwcesign) [inline pkg/detector/preemption.go:110]:
```go
defer func() {
	metrics.CountPolicyPreemption(err)
	if err != nil {
		d.EventRecorder.Eventf(resourceTemplate, corev1.EventTypeWarning, events.EventReasonPreemptPolicyFailed,
			"Propagation policy(%s/%s) failed to preempt propagation policy(%s/%s): %v", policy.Namespace, policy.Name, claimedPolicyNamespace, claimedPolicyName, err)
		return
	}
	d.EventRecorder.Eventf(resourceTemplate, corev1.EventTypeNormal, events.EventReasonPreemptPolicySucceed,
		"Propagation policy(%s/%s) preempted propagation policy(%s/%s) successfully", policy.Namespace, policy.Name, claimedPolicyNamespace, claimedPolicyName)
}()
```

File: pkg/detector/preemption.go [hunk 2]
```diff
@@ -115,13 +129,24 @@ func (d *ResourceDetector) preemptPropagationPolicy(resourceTemplate *unstructur
...
+		metrics.CountPolicyPreemption(err)
+		if err != nil {
+			d.EventRecorder.Eventf(resourceTemplate, corev1.EventTypeWarning, events.EventReasonPreemptPolicyFailed,
+				"Propagation policy(%s/%s) failed to preempt cluster propagation policy(%s): %v", policy.Namespace, policy.Name, claimedPolicyName, err)
+		} else {
```
Comment (jwcesign) [inline pkg/detector/preemption.go:143]: ditto

**Review by jwcesign — COMMENTED** [review]
/lgtm

**Comment by whitewindmills** [comment]
ptal @jwcesign @RainbowMango @chaunceyjiang @Poor12

**Comment by Poor12** [comment]
/lgtm
Please remember to update the docs about metrics and event.

**Comment by whitewindmills** [comment]
cc @RainbowMango to approve

**Review by RainbowMango — APPROVED** [review]
/approve
Thanks.

**Pull request merged**